sdk/core/allocator/alloc.h

```diff
@@ -1160,26 +1250,135 @@ class MState
+				heap.address() = ptr.address();
+				auto chunk     = MChunkHeader::from_body(heap);
+				// We know this isn't in the hazard lists (we just checked!) so
+				// free it without doing any hazard pointer checks checks.
+				mspace_free(*chunk, ptr.length(), true);
+			}
+		}
+		hazardQuarantineSize = insert;
+		return foundSkippedValue;
+	}
+
 	/**
 	 * Free a chunk.  The `bodySize` parameter specifies the size of the
 	 * allocated space that must be zeroed.  This must be calculated by the
 	 * caller and so is provided here to avoid recalculating it.
 	 */
-	int mspace_free(MChunkHeader &chunk, size_t bodySize)
+	int mspace_free(MChunkHeader &chunk,
+	                size_t        bodySize,
+	                bool          skipHazardCheck = false)
 	{
 		// Expand the bounds of the freed object to the whole heap and set the
 		// address that we're looking at to the base of the requested
 		// capability.
-		CHERI::Capability mem{chunk.body()};
+		Capability mem{chunk.body()};
 
-		/*
-		 * Paint the shadow bitmap and then zero the memory contents.  These
-		 * must happen in this order, since the allocator is running with
-		 * interrupts enabled.  Were we to zero and then paint, there would be a
-		 * window in which preemption could allow a store through a copy of
-		 * the user capability (or its progeny) that undid our work of zeroing!
-		 */
-		revoker.shadow_paint_range<true>(mem.address(), chunk.cell_next());
+		bool isDoubleFree = false;
+
+		if (__predict_false(skipHazardCheck))
+		{
+			// Paint before zeroing, see comment on the `else` code path.
+			revoker.shadow_paint_range<true>(mem.address(), chunk.cell_next());
+		}
+		else
+		{
+			Capability bounded{mem};
+			bounded.bounds() = bodySize;
+
+			auto guard = hazard_list_begin();
+
+			// Free any objects whose lifetimes were extended by hazards,
+			// skipping freeing this one to avoid a double free.
+			// Track if we're freeing an object that has already been freed but
+			// is kept alive by a hazard pointer.
+			isDoubleFree = hazard_pointers_recheck(bounded);
+
+			// If this object is on a live hazard list, capture it and don't
+			// free yet.
+			if (hazard_pointer_check(bounded))
+			{
+				hazard_quarantine_add(bounded);
+				return isDoubleFree ? -EINVAL : 0;
+			}
+			if (!bounded.is_valid())
+			{
+				return -EINVAL;
+			}
+
+			/*
+			 * Paint the shadow bitmap with the hazard epoch odd.  This must
+			 * happen before we increment the epoch so that any threads that saw
+			 * that they are waiting for us to walk the epoch list will block
+			 * until we're done.
+			 *
+			 * It must also happen before zeroing because the allocator is
+			 * running with interrupts enabled.  Were we to zero and then paint,
+			 * there would be a window in which preemption could allow a store
+			 * through a copy of the user capability (or its progeny) that undid
+			 * our work of zeroing!
+			 */
+			revoker.shadow_paint_range<true>(mem.address(), chunk.cell_next());
```

Comment:
Perhaps a comment of `guard` going out of scope after this line, just so there's _something_ written that makes it look like this can't be CSE'd with the then branch above.

Reply:
Does the comment above not say this?  It starts with 'Paint the shadow bitmap with the hazard epoch odd'.  I've added a comment to the guard.

Reply:
Sorry, I'd meant a comment _after_ the call to `shadow_paint_range` so that someone like future-me, reading upwards from the ends of the then and `else` scopes wouldn't forget that `guard` going out of scope was important.